src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -327,6 +410,148 @@ public void ConnectHeaders_SupplyHeaders_ToDownloadArraySizeSameAsNumberOfHeader
+            ChainedHeader chainTip = testContext.ExtendAChain(7);
+            chainedHeaderTree.Initialize(chainTip, true);
+
+            // Chain A is extended by Peer 1:
+            // 1a - 2a - 3a - 4a - 5a - 6a - 7a - 8a - 9a - 10a - 11a - 12a - 13a - 14a
+            ChainedHeader peer1ChainTip = testContext.ExtendAChain(6, chainTip);
+            List<BlockHeader> listOfPeer1NewHeaders = testContext.ChainedHeaderToList(peer1ChainTip, 6);
+            List<BlockHeader> peer1NewHeaders = listOfPeer1NewHeaders.GetRange(0, 4);
+
+            // 4 headers are processed 2 are unprocessed.
+            ConnectNewHeadersResult connectedNewHeadersResult = chainedHeaderTree.ConnectNewHeaders(1, peer1NewHeaders);
+            Assert.Equal(2, peer1ChainTip.Height - connectedNewHeadersResult.Consumed.Height);
+
+            // Remaining headers are presented.
+            connectedNewHeadersResult = chainedHeaderTree.ConnectNewHeaders(1, listOfPeer1NewHeaders.GetRange(4, 2));
+            
+            // All headers are now consumed.
+            Assert.Equal(connectedNewHeadersResult.Consumed.HashBlock, peer1ChainTip.HashBlock);
+
+            // Peer 2 presents headers that that are already in the tree: 11a - 12a - 13a - 14a
+            // As well as some that are not in it yet: 15a - 16a
+            ChainedHeader peer2ChainTip = testContext.ExtendAChain(2, peer1ChainTip); 
+            List<BlockHeader> listOfNewAndExistingHeaders = testContext.ChainedHeaderToList(peer2ChainTip, 6);
+
+            connectedNewHeadersResult = chainedHeaderTree.ConnectNewHeaders(2, listOfNewAndExistingHeaders);
+            Assert.True(connectedNewHeadersResult.Consumed.Height == peer2ChainTip.Height);
+
+            // Peer 3 presents a list of headers that are already in the tree
+            // and the rest of the headers from a new fork to the tree.
+            // 13a - 14a - 15a - 16a
+            //     - 14b - 15b - 16b
+            ChainedHeader peer3ChainTip = testContext.ExtendAChain(2, peer2ChainTip.Previous.Previous.Previous);
+            List<BlockHeader> listOfPeer3Headers = testContext.ChainedHeaderToList(peer3ChainTip, 5);
+            connectedNewHeadersResult = chainedHeaderTree.ConnectNewHeaders(3, listOfPeer3Headers);
+
+            Assert.True(connectedNewHeadersResult.Consumed.Height == peer3ChainTip.Height);
+            
+            List<BlockHeader> peerOneChainedHeaderList = testContext.ChainedHeaderToList(peer1ChainTip, 7);
+            List<BlockHeader> peerTwoChainedHeaderList = testContext.ChainedHeaderToList(peer2ChainTip, 7);
+            
+            // Submit headers that are all already in the tree:
+            // Peer 3 supplies all headers from Peer 1.
+            connectedNewHeadersResult = chainedHeaderTree.ConnectNewHeaders(3, peerOneChainedHeaderList);
+            Assert.True(connectedNewHeadersResult.Consumed.Height == chainedHeaderTree.GetPeerTipChainedHeaderByPeerId(3).Height);
```

Comment:
As above, especially if there are more branches, the height is not unique

Also below it is the same.

Reply:
Fixed.